sql: Internal sessions exempt from disallow_full_table_scans setting

spilchen opened the pull request:
Previously, enabling the cluster setting sql.defaults.disallow_full_table_scans.enabled prevented the creation of indexes due to the internal table scans required during the process. This change ensures that the setting is bypassed for internal operations, while still applying to user queries.

Epic: None
Closes: #137404
Release note (bug fix): Internal scans are now exempt from the sql.defaults.disallow_full_table_scans.enabled setting, allowing index creation even when the cluster setting is active.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137681)
<!-- Reviewable:end -->

rafiss reviewed (approved):
lgtm!

spilchen committed:
sql: Internal sessions exempt from disallow_full_table_scans setting

Previously, enabling the cluster setting
sql.defaults.disallow_full_table_scans.enabled prevented the creation of
indexes due to the internal table scans required during the process.
This change ensures that the setting is bypassed for internal
operations, while still applying to user queries.

Epic: None
Closes: #137404
Release note (bug fix): Internal scans are now exempt from the
sql.defaults.disallow_full_table_scans.enabled setting, allowing index
creation even when the cluster setting is active.

spilchen commented on pkg/sql/logictest/testdata/logic_test/create_index:
```diff
@@ -440,6 +440,27 @@ CREATE INVERTED INDEX ON opclasses(c DESC)
 statement ok
 CREATE INVERTED INDEX ON opclasses(a DESC, c)
 
+# Regression test for GH issue #137404
```
Good suggestion. Thanks.
<!-- Reviewable comment -OEQM8612co3NhJnpakM:b-r1cp1r -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137681)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

michae2 reviewed (approved):
LGTM, but I thought [this check for executorType](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/conn_executor_exec.go#L3195) meant that internal executor was already exempt from disallow_full_table_scans?

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137681#-:-OEQh_Xi2bEiDKCj7JNH:b-yqdp5f)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
We never reach line 3195:
https://github.com/cockroachdb/cockroach/blob/f26c138ced4917a5e2af8f548025fac1e8f47bed/pkg/sql/conn_executor_exec.go#L3194-L3195

The previous condition is false for CREATE INDEX. The only flags set are:
- planFlagIsDDL
- planFlagContainsMutation

Should I revert this fix and work on getting the `planFlagContainsFullTableScan` flag set for CREATE INDEX instead?

rafiss commented:
> Should I revert this fix and work on getting the planFlagContainsFullTableScan flag set for CREATE INDEX instead?

I'd prefer to leave to the SQL Queries team to decide how to tackle that (cc @michae2). In the meantime, I think the fix you've made here is a robust way of addressing the issue, and it uses the same pattern we have for statement_timeout so i don't think we need to change it.

michae2 commented:
> We never reach line 3195:
> 
> https://github.com/cockroachdb/cockroach/blob/f26c138ced4917a5e2af8f548025fac1e8f47bed/pkg/sql/conn_executor_exec.go#L3194-L3195
> 
> The previous condition is false for CREATE INDEX. The only flags set are:
> 
> * planFlagIsDDL
> * planFlagContainsMutation
> 
> Should I revert this fix and work on getting the `planFlagContainsFullTableScan` flag set for CREATE INDEX instead?

If we never reach line 3195, then how is `disallow_full_table_scans` preventing the creation of indexes?

spilchen commented:
Without this fix, CREATE INDEX fails at this point:
https://github.com/cockroachdb/cockroach/blob/a2db3126b8ff10e351c6d5748555d72cf712f1ce/pkg/sql/opt/exec/execbuilder/relational.go#L604-L605

It's driving a full index scan of the form `SELECT count(1) FROM [%d AS t]@[%d]` from here:
https://github.com/cockroachdb/cockroach/blob/e6a5104f85484b185550d6c4c2e8c588fa3d6d47/pkg/sql/backfill.go#L2104-L2111

In `conn_executor_exec.go`, it's really close to the `DisallowFullTableScan` check you mentioned. It's at line 3187 here. 
https://github.com/cockroachdb/cockroach/blob/503d914d99b50ed96d66f7fd01247e588dfb5592/pkg/sql/conn_executor_exec.go#L3187-L3195

I don't believe the statements in `conn_executor_exec.go` can be reordered, as the logic depends on the flags from the planner, which are still being built.

michae2 commented:
I see, thanks. I didn't realized we also checked disallow_full_table_scans in execbuild when the index is forced. And that check doesn't look at the executorType.

I'm going to open a follow-up PR removing the executorType part, since this fix is cleaner and the executorType part doesn't work in execbuild anyway.
